### PR TITLE
refactor: reject impossible casts

### PR DESCRIPTION
**Pull request by jaschdoc**
Closes https://github.com/flix/flix/issues/4762

### CODE REVIEW

**Comment by jaschdoc** [comment]
Do we allow casting from `String` to `Char`?

**Comment by jaschdoc** [comment]
Also, I think this might break lots of casts in the library. Should it maybe be less restrictive?

**Comment by magnus-madsen** [comment]
Impossible should only be truely impossible, e.g. Int32 to String.

**Comment by magnus-madsen** [comment]
> Do we allow casting from `String` to `Char`?

No, that always fails.

**Comment by magnus-madsen** [comment]
I think you should keep it extremely simple:

- A primitive can never be cast to a reference type (or vise versa).
- A non-Boolean primitive cannot be cast to a boolean (or vise versa).

You never reject anything with type variables or the like. The purpose here is just to catch common really stupid things that go wrong. Not to do anything fancy.

**Comment by jaschdoc** [comment]
> A non-Boolean primitive cannot be cast to a boolean (or vise versa).

So `char` for instance is allowed to be cast to `Int32` to read it as a 32-bit number?

**Comment by magnus-madsen** [comment]
The kava spec risallpws that.

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 1]
```diff
@@ -316,6 +317,52 @@ object Safety {
...
+      Type.Unit :: Type.Bool :: Type.Char ::
+        Type.Float32 :: Type.Float64 :: Type.Int8 ::
+        Type.Int16 :: Type.Int32 :: Type.Int64 ::
+        Type.Str :: Type.BigInt :: Type.BigDecimal :: Nil
+    }.toSet
+
+    (tpe1, tpe2) match {
+
+      // Reference types
+      case (t1, Some(Type.Cst(TypeConstructor.Ref, _))) if primitives.contains(t1) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:342]: Ah, sorry, but reference types I did not mean ` Ref`, but anything that is an object on the JVM.

**Comment by magnus-madsen** [comment]
In other words, its really this:

A primitive can never be cast to a non-primitive type (or vise versa).
A non-Boolean primitive cannot be cast to a boolean (or vise versa).

So the logic would probably be:

if (one of the primitive types, but not bool) then can cast to one of the primitive types (but not bool).

**Comment by jaschdoc** [comment]
Alright, this version only handle reference types and boolean primitives. No special rules wrt. Int32 -> Int64 or anything of the sort between primitive types except for booleans.

**Pull request merged**